### PR TITLE
Tvs/typescript react18

### PR DESCRIPTION
**Pull request by tonypls**
## Issue

Convert the web codebase to use TypeScript and react 18 🎉 

## Description

Change the codebase to typescript to make the lives of contributors better and bring more structure to the codebase as the project scales.

TS will bring us: 
Optional typing 
Great tooling like IntelliSense :)

React 18 will bring us:
Access to modern libraries
New apis and hooks
Concurrent rendering
Strict mode

Todos: 
- Refactor deprecated "ReactDOM.render", our version of react will operate at 17 until this is completed
- Add types where logical through the app
- Work towards a no "any" codebase 🏆

### CODE REVIEW

**Comment by VIKTORVAV99** [comment]
Just some quick thoughts and questions:
- Wasn't React 18 blocked by `react-map-gl'?
- If we need to update `react-map-gl` can we switch to [maplibre](https://github.com/maplibre/maplibre-gl-js)? (it's a open source fork of mapbox)
- Why are we importing the entire `lodash` library? 
- The babel config and possibly webpack config need to be updated as well.

**Comment by tonypls** [comment]
> Just some quick thoughts and questions:
> 
> * Wasn't React 18 blocked by `react-map-gl`?
> * If we need to update `react-map-gl` can we switch to [maplibre](https://github.com/maplibre/maplibre-gl-js)? (it's a open source fork of mapbox)
> * Why are we importing the entire `lodash` library?
> * The babel config and possibly webpack config need to be updated as well.

Hey Viktor!

Thanks for the feedback.

* I think there is compatibility with `react-map-gl` and react 18 but I will get react 18 running and verify this, if not maplibre looks nice
* Imported lodash while I was getting things up and running, will remove it to keep the bundle size down :)
* Will look into the configs today

**Comment by VIKTORVAV99** [comment]
> > Just some quick thoughts and questions:
> > 
> > * Wasn't React 18 blocked by `react-map-gl`?
> > * If we need to update `react-map-gl` can we switch to [maplibre](https://github.com/maplibre/maplibre-gl-js)? (it's a open source fork of mapbox)
> > * Why are we importing the entire `lodash` library?
> > * The babel config and possibly webpack config need to be updated as well.
> 
> Hey Viktor!
> 
> Thanks for the feedback.
> 
> * I think there is compatibility with `react-map-gl` and react 18 but I will get react 18 running and verify this, if not maplibre looks nice
> * Imported lodash while I was getting things up and running, will remove it to keep the bundle size down :)
> * Will look into the configs today
> 
> 

Nice!
Looks like you got things under control then!

Just FYI maplibre is a replacement for mapbox not react-map-gl.
(the newer versions of mapbox require a API key and charges for each render so it would be great if we could move away from that.)

**Comment by Kongkille** [comment]
> > > Just some quick thoughts and questions:
> > > 
> > > * Wasn't React 18 blocked by `react-map-gl`?
> > > * If we need to update `react-map-gl` can we switch to [maplibre](https://github.com/maplibre/maplibre-gl-js)? (it's a open source fork of mapbox)
> > > * Why are we importing the entire `lodash` library?
> > > * The babel config and possibly webpack config need to be updated as well.
> > 
> > 
> > Hey Viktor!
> > Thanks for the feedback.
> > 
> > * I think there is compatibility with `react-map-gl` and react 18 but I will get react 18 running and verify this, if not maplibre looks nice
> > * Imported lodash while I was getting things up and running, will remove it to keep the bundle size down :)
> > * Will look into the configs today
> 
> Nice! Looks like you got things under control then!
> 
> Just FYI maplibre is a replacement for mapbox not react-map-gl. (the newer versions of mapbox require a API key and charges for each render so it would be great if we could move away from that.)

If we end up switching away from Mapbox then we also need to ensure Cordova is able to output native apps with the alternative

**Comment by VIKTORVAV99** [comment]
> > > > Just some quick thoughts and questions:
> > > > 
> > > > * Wasn't React 18 blocked by `react-map-gl`?
> > > > * If we need to update `react-map-gl` can we switch to [maplibre](https://github.com/maplibre/maplibre-gl-js)? (it's a open source fork of mapbox)
> > > > * Why are we importing the entire `lodash` library?
> > > > * The babel config and possibly webpack config need to be updated as well.
> > > 
> > > 
> > > Hey Viktor!
> > > Thanks for the feedback.
> > > 
> > > * I think there is compatibility with `react-map-gl` and react 18 but I will get react 18 running and verify this, if not maplibre looks nice
> > > * Imported lodash while I was getting things up and running, will remove it to keep the bundle size down :)
> > > * Will look into the configs today
> > 
> > Nice! Looks like you got things under control then!
> > 
> > Just FYI maplibre is a replacement for mapbox not react-map-gl. (the newer versions of mapbox require a API key and charges for each render so it would be great if we could move away from that.)
> 
> If we end up switching away from Mapbox then we also need to ensure Cordova is able to output native apps with the alternative

Since it's a fork of the current mapbox version we use I don't think it will be a major issue. It's still JS after all and Babel will transform code where it's needed.

File: web/cypress/integration/countrypanel.spec.ts [hunk 1]
```diff
@@ -1,13 +1,13 @@
 describe('Country Panel', () => {
```
Comment (tonypls) [inline web/cypress/integration/countrypanel.spec.ts:1]: I think we can just ignore cypress and leave it as JS

**Comment by tonypls** [comment]
We should replace the outdated libraries mentioned here:
https://github.com/electricitymap/electricitymap-contrib/issues/4143#issue-1239484001

[react-motion](https://github.com/chenglou/react-motion)

[react-transition-group](https://github.com/reactjs/react-transition-group)

They might not be compatible but from my testing both work on Chrome

**Comment by VIKTORVAV99** [comment]
> We should replace the outdated libraries mentioned here:
> https://github.com/electricitymap/electricitymap-contrib/issues/4143#issue-1239484001
> 
> [react-motion](https://github.com/chenglou/react-motion)
> 
> [react-transition-group](https://github.com/reactjs/react-transition-group)
> 
> They might not be compatible but from my testing both work on Chrome

I (and Mads) got started on this bit we found no good alternative, we also talked about using a viz library for all the graphs but IDK what came out of that.

Have you (the team) discussed this?

**Comment by tonypls** [comment]
I've merged master but this needs testing and removing of /* eslint-disable */ especially around the country panel

**Comment by VIKTORVAV99** [comment]
@tonypls would you mind if I clean up some of the merge conflicts and begin to test this?

Will probably take a while as it's a massive PR but we have to start somewhere.

**Pull request closed without merging**